### ISSUE TITLE
[PATCH] proper parameter formatting better default arg handling in :cpp:function

### ISSUE THREAD

**Issue by shimizukawa** [sphinx-doc/sphinx#681]
Hello,

This is a patch for sphinx/domains/cpp.py from Sphinx 1.0.7 that adds :param: directive support to :cpp:function: and also handles the default parameters containing parentheses, e.g. :cpp:function: void MyClass::foo(const string& name=string());

Regards,
Vadim

```
--- /Users/vp/Downloads/Sphinx-1.0.7/sphinx/domains/cpp.py  2011-01-08 18:05:59.000000000 +0300
+++ /Users/vp/Downloads/Sphinx-1.0.7/sphinx/domains/cpp2.py 2011-04-19 16:10:06.000000000 +0400
@@ -21,7 +21,7 @@
 from sphinx.directives import ObjectDescription
 from sphinx.util.nodes import make_refnode
 from sphinx.util.compat import Directive
-
+from sphinx.util.docfields import Field, TypedField

 _identifier_re = re.compile(r'(~?\b[a-zA-Z_][a-zA-Z0-9_]*)\b')
 _whitespace_re = re.compile(r'\s+(?u)')
@@ -658,17 +658,33 @@
         self.skip_ws()
         if self.match(_string_re):
             return self.matched_text
-        idx1 = self.definition.find(',', self.pos)
-        idx2 = self.definition.find(')', self.pos)
-        if idx1 < 0:
-            idx = idx2
-        elif idx2 < 0:
-            idx = idx1
-        else:
-            idx = min(idx1, idx2)
-        if idx < 0:
-            self.fail('unexpected end in default expression')
-        rv = self.definition[self.pos:idx]
+        paren_stack_depth = 0
+        max_pos = len(self.definition)
+        rv_start = self.pos
+        while 1:
+            idx0 = self.definition.find('(', self.pos)
+            idx1 = self.definition.find(',', self.pos)
+            idx2 = self.definition.find(')', self.pos)
+            if idx0 < 0:
+                idx0 = max_pos
+            if idx1 < 0:
+                idx1 = max_pos
+            if idx2 < 0:
+                idx2 = max_pos
+            idx = min(idx0, idx1, idx2)
+            if idx >= max_pos:
+                self.fail('unexpected end in default expression')
+            if idx == idx0:
+                paren_stack_depth += 1
+            elif idx == idx2:
+                paren_stack_depth -= 1
+                if paren_stack_depth < 0:
+                    break
+            elif paren_stack_depth == 0:
+                break
+            self.pos = idx+1
+            
+        rv = self.definition[rv_start:idx]
         self.pos = idx
         return rv

@@ -788,6 +804,16 @@
 class CPPObject(ObjectDescription):
     """Description of a C++ language object."""

+    doc_field_types = [
+        TypedField('parameter', label=l_('Parameters'),
+                   names=('param', 'parameter', 'arg', 'argument'),
+                   typerolename='type', typenames=('type',)),
+        Field('returnvalue', label=l_('Returns'), has_arg=False,
+              names=('returns', 'return')),
+        Field('returntype', label=l_('Return type'), has_arg=False,
+              names=('rtype',)),
+    ]
+
     def attach_name(self, node, name):
         owner, name = name.split_owner()
         varname = unicode(name)
```

---
- Bitbucket: https://bitbucket.org/birkenfeld/sphinx/issue/681
- Originally reported by: Anonymous
- Originally created at: 2011-04-19T14:16:38.642

**Comment by shimizukawa** [comment]
_From Jeroen Keiren on 2012-01-03 15:21:29+00:00_

I second that this is a useful modification. It applies mostly cleanly to the current development version, and it would be great to have this change in the next release.

The patch fixes all issues of functions with default parameters that I have run into so far.

**Comment by shimizukawa** [comment]
_From Jim Naslund on 2012-11-10 16:47:36+00:00_

This works for me too. I've rebased just the default argument parsing fix against the tip and added a test in the attached patch. I can create a pull request if that is preferred.

**Comment by shimizukawa** [comment]
_From Jon Waltman on 2012-11-11 08:18:06+00:00_

Closes #681: Allow nested parentheses in C++ signatures (patch by Vadim and Jim Naslund)

→ <<cset c134b218b743206697ce1843b2a529d7ec526d41>>
